pages/papers/attention-is-all-you-need.mdx

```diff
@@ -1,3 +1,223 @@
+
+There are two important things to note here
+
+1. We scale $QK^T$ by $\sqrt{d_k}$ because of numerical instability. This is because the product of two matrices has a significantly greater variance and tends to lead to large values being generated. Scaling down allows for a smaller value to be used when computing the softmax, resulting in better results
+2. The Dot Product is chosen because it is much faster and more space-efficient in practice, since it can be implemented using highly optimised matrix multiplication code.
+
+The model tunes the values of Q and K so that it learns how to assign the proper weightage to each token for a given sequence. We can intuitively understand the Query matrix as a representation of a sequence seeking a match while the Key matrix is a representation of a sequence's meaning.
+
+![transformer-softmax](/transformer-softmax.png)
+
+#### Multi-Head Attention
+
+![MHA-Formula](/MHA-attention-formula.png)
+
+Multi-Head attention takes advantage of the Scaled Dot Product Attention mechanism. While the computation complexity is the same as that of a Single-head attention block, the authors find that it performs better.
+
+This is because it allows the model to jointly attend to information from different representation subspaces at different positions. In the paper, they used `8` different heads.
+
+![MHA-Attention](/mha-attention-visualisation.png)
+
+Note that at the end when we derive our H matrix, it has dimensions $(n, d_v)$ but with the help of $w^o$, we’re able to scale our output back to $d_{model}$.
+
+
+
+#### Decoder Masks
+
+In the decoder, we apply a mask to our attention matrix which only allows softmax for each row to be calculated using only the corresponding token for that row and the ones before it. This prevents our model from being able to make predictions using future states.
+
+#### Cross Attention vs Self-Attention
+
+We can see here that our model uses two forms of attention. 
+
+![Transformer Attention](/transformer-attentions-type.png)
+
+The first is Self-Attention, where the Q, K and V matrices are derived from the same input vector (As seen above). However, we also have Cross-Attention, where the Q,K and V matrices are not. 
+
+In the example above, our decoder Multi-Head attention block derives Q and K from the encoder’s output and V from the output of the previous Multi-Head Attention block. 
+
+> Note that the same encoder output is passed to every single decoder block. Hence, information is encoded once and then cached for every subsequent inference or training call
+>
+
+## Encoders and Decoders
+
+> Most transformer models nowadays are decoder only because decoders tend to shine more at text generation work. Encoders helps to bridge the gap between sequences that exist in two separate embedding spaces - which is where encoder-decoder architectures tend to shine. 
```

Comment:
Bringing in some additional stuff here

Reply:
consider to add https://web.stanford.edu/class/archive/cs/cs224n/cs224n.1224/slides/cs224n-2022-lecture10-pretraining.pdf ref. slide 28 on examples